src/tsd/QueryRpc.java

```diff
@@ -115,14 +122,45 @@ private void handleQuery(final TSDB tsdb, final HttpQuery query) {
       throw new BadRequestException(HttpResponseStatus.BAD_REQUEST, 
           e.getMessage(), data_query.toString(), e);
     }
-    
+
     Query[] tsdbqueries;
     try {
       tsdbqueries = data_query.buildQueries(tsdb);
     } catch(NoSuchUniqueName ex) {
       throw new BadRequestException(ex);
     }
+
     final int nqueries = tsdbqueries.length;
+    if (query.method() == HttpMethod.DELETE) {
+      final ArrayList<Deferred<Object>> deferredsDelete =
+        new ArrayList<Deferred<Object>>(nqueries);
+      final byte[] table = tsdb.getConfig().getString("tsd.storage.hbase.data_table").getBytes();
+      final HBaseClient client = tsdb.getClient();
+
+      LOG.info("deleting from table=" + Arrays.toString(table));
+      try {
+        for (final Query q : tsdbqueries) {
+          q.setEndTime(Integer.MAX_VALUE);
+          final Scanner scanner = Internal.getScanner(q);
```

Comment:
Instead of scanning HBase twice, you should pass a delete flag to the scanners in TsdbQuery and SaltScanner (Put branch) to perform the scan asynchronously and once instead of twice. This will block the thread every time we wait for HBase to fetch the next chunk of data, reducing the # of queries we can handle per second.